include/number_types.hpp

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include "chainblocks.h"
+#include "common_types.hpp"
+#include <vector>
+#include <map>
+#include <cassert>
+
+enum class NumberType {
+  Float32,
+  Float64,
+  UInt8,
+  Int8,
+  Int16,
+  Int32,
+  Int64,
+  Invalid
+};
+
+std::map<CBType, NumberType> getCBTypeToNumberTypeMap();
+const std::vector<NumberType> &getCBTypeToNumberTypeArrayMap();
+
+inline NumberType cbTypeToNumberType(CBType type) {
+  auto &mapping = getCBTypeToNumberTypeArrayMap();
+  cbassert((size_t)type < mapping.size());
+  return mapping[(size_t)type];
+}
+
+struct NumberConvertOutOfRangeEx {
```

Comment:
I don't know about the best practice in C++, but in .NET exception names are nouns not verbs.
There fore I would have named it `NumberConversionOutOfRangeEx`.